test/e2e/storage/testsuites/multivolume.go

```diff
@@ -0,0 +1,468 @@
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+type multiVolumeTestSuite struct {
+	tsInfo TestSuiteInfo
+}
+
+var _ TestSuite = &multiVolumeTestSuite{}
+
+// InitMultiVolumeTestSuite returns multiVolumeTestSuite that implements TestSuite interface
+func InitMultiVolumeTestSuite() TestSuite {
+	return &multiVolumeTestSuite{
+		tsInfo: TestSuiteInfo{
+			name: "multiVolume",
+			testPatterns: []testpatterns.TestPattern{
+				testpatterns.FsVolModePreprovisionedPV,
+				testpatterns.FsVolModeDynamicPV,
+				testpatterns.BlockVolModePreprovisionedPV,
+				testpatterns.BlockVolModeDynamicPV,
+			},
+		},
+	}
+}
+
+func (t *multiVolumeTestSuite) getTestSuiteInfo() TestSuiteInfo {
+	return t.tsInfo
+}
+
+func (t *multiVolumeTestSuite) defineTests(driver TestDriver, pattern testpatterns.TestPattern) {
+	type local struct {
+		config      *PerTestConfig
+		testCleanup func()
+
+		cs        clientset.Interface
+		ns        *v1.Namespace
+		driver    TestDriver
+		resources []*genericVolumeTestResource
+	}
+	var (
+		dInfo = driver.GetDriverInfo()
+		l     local
+	)
+
+	BeforeEach(func() {
+		// Check preconditions.
+		if pattern.VolMode == v1.PersistentVolumeBlock && !dInfo.Capabilities[CapBlock] {
+			framework.Skipf("Driver %s doesn't support %v -- skipping", dInfo.Name, pattern.VolMode)
+		}
+	})
+
+	// This intentionally comes after checking the preconditions because it
+	// registers its own BeforeEach which creates the namespace. Beware that it
+	// also registers an AfterEach which renders f unusable. Any code using
+	// f must run inside an It or Context callback.
+	f := framework.NewDefaultFramework("multivolume")
+
+	init := func() {
+		l = local{}
+		l.ns = f.Namespace
+		l.cs = f.ClientSet
+		l.driver = driver
+
+		// Now do the more expensive test initialization.
+		l.config, l.testCleanup = driver.PrepareTest(f)
+	}
+
+	cleanup := func() {
+		for _, resource := range l.resources {
+			resource.cleanupResource()
+		}
+
+		if l.testCleanup != nil {
+			l.testCleanup()
+			l.testCleanup = nil
+		}
+	}
+
+	// This tests below configuration:
+	//          [pod1]                            same node       [pod2]
+	//      [   node1   ]                           ==>        [   node1   ]
+	//          /    \      <- same volume mode                   /    \
+	//   [volume1]  [volume2]                              [volume1]  [volume2]
+	It("should access to two volumes with the same volume mode and retain data across pod recreation on the same node", func() {
+		// Currently, multiple volumes are not generally available for pre-provisoined volume,
+		// because containerized storage servers, such as iSCSI and rbd, are just returning
+		// a static volume inside container, not actually creating a new volume per request.
+		if pattern.VolType == testpatterns.PreprovisionedPV {
+			framework.Skipf("This test doesn't work with pre-provisioned volume -- skipping")
+		}
+
+		init()
```

Comment:
every test case is doing init/cleanup.  Can this go in Before/AfterEach instead?

Reply:
Current code is a bit redundant, but it give us a chance to skip tests before driver initialization, per test basis. (Some test can be skipped before the information of driver, but others won't.)
If we prioritize readability over performance, we will be able to choose moving it to Before/AfterEach.

Reply:
I guess for now it's fine, but I eventually want to move to a model where we bring up the driver during test cluster initialization instead of during the test case.